libraries/ESP8266WiFi/src/ESP8266WiFiSTA.cpp

```diff
@@ -197,55 +197,57 @@ wl_status_t ESP8266WiFiSTAClass::begin() {
     return status();
 }
 
-static void
-swap(IPAddress &lhs, IPAddress &rhs)
-{
-  IPAddress tmp = lhs;
-  lhs = rhs;
-  rhs = tmp;
-}
-
-
 /**
  * Change IP configuration settings disabling the dhcp client
  * @param local_ip   Static ip configuration
  * @param gateway    Static gateway configuration
  * @param subnet     Static Subnet mask
  * @param dns1       Static DNS server 1
  * @param dns2       Static DNS server 2
+ * or
+ * @param local_ip   Static ip configuration
+ * @param dns1       Static DNS server
+ * @param gateway    Static gateway configuration
+ * @param subnet     Static Subnet mask
  */
-bool ESP8266WiFiSTAClass::config(IPAddress local_ip, IPAddress gateway, IPAddress subnet, IPAddress dns1, IPAddress dns2) {
+bool ESP8266WiFiSTAClass::config(IPAddress local_ip, IPAddress arg1, IPAddress arg2, IPAddress arg3, IPAddress dns2) {
 
     if(!WiFi.enableSTA(true)) {
         return false;
     }
 
-    //Arduino has a different arg order: ip, dns, gateway, subnet. To allow compatibility, check first octet of 3rd arg. If 255, interpret as ESP order, otherwise Arduino order.
+    //ESP argument order is: ip, gateway, subnet, dns1
+    //Arduino arg order is:  ip, dns, gateway, subnet.
+
+    //first, check whether dhcp should be used, which is when ip == 0 && gateway == 0 && subnet == 0.
+    bool espOrderUseDHCP = (local_ip == 0U && arg1 == 0U && arg2 == 0U);
+    bool arduinoOrderUseDHCP = (local_ip == 0U && arg2 == 0U && arg3 == 0);
+    if (espOrderUseDHCP || arduinoOrderUseDHCP) {
+        _useStaticIp = false;
+        wifi_station_dhcpc_start();
+        return true;
+    }
+
+
+    //To allow compatibility, check first octet of 3rd arg. If 255, interpret as ESP order, otherwise Arduino order.
+    IPAddress gateway = arg1;
+    IPAddress subnet = arg2;
+    IPAddress dns1 = arg3;
+
     if(subnet[0] != 255)

```

Comment:
We could here check if ip and gateway are in the same network (ip & netmask) == (gateway & netmask) which gives :
- arduino if: (local_ip & arg3) == (arg2 & arg3)
- esp if: (local_ip & arg2) == (arg1 & arg2)

Reply:
Agreed about the check, but I would add it later....